### PR TITLE
ENH: remove duplicate function call

### PR DESCRIPTION
**Pull request by WRKampi**
`func` has already been evaluated at `xb` in `get_bracket_info()`, called a few lines above (line 2350).

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
If this is indeed correct, looks like at least one docstring will need the number of function evaluations to be reduced accordingly, if I'm reading the CI failure correctly.

**Comment by WRKampi** [comment]
@andyfaff 
I seem to have removed the request for your review by accident. I don't see how I can rerequest it, other than by sending you this message.

**Comment by mdhaber** [comment]
@WRKampi Could you show that `fb` and `func(*((x,) + self.args))` are guaranteed to be identical, and also give a practical example that shows a substantial improvement by making the change?  I think that would help speed up review. It is not always fun to trace the many paths that code can take, and some may not want to do so unless the benefits outweigh the risks of such a change.

**Comment by WRKampi** [comment]
Thanks for your comment. 

Regarding risk/benefit, I think that preventing double work is the low hanging fruit in such optimization routines. Even though the benefit is only 1 function call, this could be a CFD simulation that takes a day to run. (My personal tolerance does not exceed a minute though.)

Tracing back the code paths is no fun indeed. So let me try to give it a shot. Everything is in the `_optimize.py` file. The main question is if `xb` and `fb` are such that `fb == func(*((xb,) + self.args))` in line 2350. We need to look at `get_bracket_info()` for answers.

`get_bracket_info()` starts at line 2309. It either raises a `ValueError` (which is OK), or returns the bracket. There are 3 paths. If `len(brack) == 3`, line 2321, we can directly see that `fb = func(*((xb,) + args))` on line 2331. The other two paths call the function `bracket`. 

This function starts on line 2785. It is either called with two x values, or it uses `xa=0.0` and `xb=1.0`. The bracket function can raise a `RuntimeError` (which is OK) or return the output including `xb` and `fb`. The code of `bracket` is harder to evaluate, so let's test it instead.

```
"""test if the function `bracket` operates as advertised. Specifically, the
returned `fb` must equal `func(*((xb,) + args))`. Raising a RuntimeError is
acceptable too."""
from scipy.optimize import bracket
from numpy.random import random_sample
#print(help(bracket))

def func(x, a, b):
    return 1./(x+a) + x/b

args = (1., 2.)
N = 10_000
for i in range(N):
    xa = random_sample()
    xb = random_sample()
    try:
        xa, xb, xc, fa, fb, fc, funcalls = bracket(func, xa, xb, args)
    except RuntimeError:
        print("OK, RuntimeError")
    else:
        assert fb == func(*((xb,) + args)), "We have a problem"   # THE test!
        print(f"OK, {funcalls} calls")
```
This script prints "OK" and the number of function calls if `fb == func(*((xb,) + self.args))` and raises an `AssertionError` otherwise. I get only OKs and different numbers of function calls when I run it. It is possible to change `func` and `args` above, as long as it can be evaluated.

I hope this will speed up the review process.

**Comment by mdhaber** [comment]
In this case, I'm not entirely convinced by the randomized test. Does it take all the code paths? Could the order matter?

Maybe we can reason this out, though. I agree with what you have above, and it remains to be shown that the `fb` and `xb` returned by `bracket` is consistent with each other; i.e. `fb == func(*(xb,) + args)` (always).

There are two assignments of `fb` for which this is trivial:
https://github.com/scipy/scipy/blob/20c80c943a5e4bc557ef721845f8c73159bfa520/scipy/optimize/_optimize.py#L2841
https://github.com/scipy/scipy/blob/20c80c943a5e4bc557ef721845f8c73159bfa520/scipy/optimize/_optimize.py#L2844

The rest of the assignments are less direct. However, every time `func` is called in the function, there is consistency between the naming of the result and the argument. E.g. In `fc = func(*((xc,) + args))`, `fc` is consistent with `xc`. Also, every time an `f_` variable is assigned, its corresponding `x_` (or `w`) variable is assigned accordingly. Is this enough to be sure `fb` and `xb` are consistent at the end?

**Comment by WRKampi** [comment]
Then we'll try it the hard way. As you pointed out, to be in a consistent state, every x has a matching f: `xa` with `fa`, `xb` with `fb`, `xc` with `fc` and `w` with `fw`. If we will never enter the while loop on line 2849, we will directly exit with the correct values. It becomes more complicated if we do enter this loop.

The while loop starts with proposing a value for `w`. The corresponding `fw` is not directly calculated (for efficiency reasons, which makes me happy). If there are too many iterations in the loop the RuntimeError could be raised, which is OK. Next, there are four possible paths though the main if, elif (2x), else statements. 

The first path is: 
https://github.com/scipy/scipy/blob/c4e7f1146556f4b038bfbfc0912ac3843e844c55/scipy/optimize/_optimize.py#L2863-L2878
`fw` is directly calculated at the top. Then there are a two ways to exit the function with a few swaps. Each x-swap has the corresponding f-swap directly after it. Therefore, the output of the function will be consistent. If the function is not exitted, this code calculates a new `w` and a new `fw` and then jumps to the bottom of the while loop in a consistent state.

The second path is:
https://github.com/scipy/scipy/blob/c4e7f1146556f4b038bfbfc0912ac3843e844c55/scipy/optimize/_optimize.py#L2879-L2882
This clearly keeps `w` and `fw` matched and does nothing to the rest of the state.

The third path is:
https://github.com/scipy/scipy/blob/c4e7f1146556f4b038bfbfc0912ac3843e844c55/scipy/optimize/_optimize.py#L2883-L2893
`fw`is calculated first. The next part is a bit confusing because a new `w` is calculated in the middle of the swapping operations. Nevertheless, the 'fw'is only calculated _after_ the corresponding f-swaps are done. So the state is consistent at the end of this code block. 

The fourth path is:
https://github.com/scipy/scipy/blob/c4e7f1146556f4b038bfbfc0912ac3843e844c55/scipy/optimize/_optimize.py#L2894-L2897
A new `w` is calculated and its corresponding `fw`. Since this is an else statement, we are sure that at least one of these four paths is taken. And each of these paths result in a consistent state ( matching x,f-pairs).

The bottom of the while loop consists of more x-swaps with corresponding f-swaps, which keeps the state consistent. So the correct values are returned if the while loop is not reentered.
https://github.com/scipy/scipy/blob/c4e7f1146556f4b038bfbfc0912ac3843e844c55/scipy/optimize/_optimize.py#L2898-L2904

To me it seems that the x,f-pairs remain matched throughout the function. The part of the while loop where `w` does not match `fw` can only be exited with a RuntimeError.

**Review by mdhaber — APPROVED** [review]
Yup, that's what it looked like to me, but thanks for the thorough job documenting it. 

Let's see if @tylerjereddy (since he commented first) or someone else is willing to do an independent review. I'd rather wait for someone else to check this before we merge.

**Comment by mdhaber** [comment]
@tylerjereddy would you be willing to take a look here? It's not domain-specific - it's just checking code logic to make sure this is indeed just a refactor, not an algorithmic change.

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:2357]: @mdhaber, I'm happy with the proposed change. `get_bracket_info` is returning `fa`, `fb`, `fc` that correspond to `xa`, `xb`, `xc`. So I don't think this function evaluation needs to happen again. If `bracket` isn't returning `f_i` corresponding to `x_i` then I think we have bigger problems.

**Review by andyfaff — APPROVED** [review]

File: scipy/optimize/_optimize.py
Comment (mdhaber) [reply scipy/optimize/_optimize.py:2357]: Yes, I suppose you're right.

**Comment by mdhaber** [comment]
Thanks @WRKampi!

**Comment by WRKampi** [comment]
Thanks @mdhaber, @andyfaff and @tylerjereddy for your help. I'm happy with my first scipy contribution.

**Pull request merged**
